### PR TITLE
Avoid data overriding & added data updater for multi-host compatibility

### PR DESCRIPTION
**Pull request by Rubenicos**
Sometimes I see this error when someone login and their data is deleted, so I fixed it.
> [EpicLevels] An error occurred executing a MySQL query: Duplicate entry 'b1532110-c636-32f8-ac4c-baab98bf3514' for key 'PRIMARY'
> com.mysql.jdbc.exceptions.jdbc4.MySQLIntegrityConstraintViolationException: Duplicate entry 'b1532110-c636-32f8-ac4c-baab98bf3514' for key 'PRIMARY'

I also fixed a problem when a player has a lot of experience.

### Changes
*  Added `DataManager#getPlayerOrCreate(UUID, Consumer)` method.
*  When player data does not exist on registeredPlayers, the data will be obtained with getPlayerOrCreate (Completely async).
*  Now experience calculator return a double value.

## Update (2022/08/05)

Mysql database connector now have an messenger to tell other servers connected at the same database when playerdata/boost need to be updated.

### Changes
*  Added DataUpdaterAbstract and DataUpdater.
*  Now `BoostManager#clearGlobalBoost` return the global boost.
*  Now commands to add/remove boosts can handle offline players data.

### CODE REVIEW

File: src/main/java/com/songoda/epiclevels/EpicLevels.java [hunk 1]
```diff
@@ -70,6 +70,7 @@ public void onPluginLoad() {
 
     @Override
     public void onPluginDisable() {
+        this.dataManager.close();
```
Comment (SpraxDev) [inline src/main/java/com/songoda/epiclevels/EpicLevels.java]: Shouldn't the DataManager be closed after the bulkUpdates have been finished? (not sure if they are async too)

File: src/main/java/com/songoda/epiclevels/database/DataManager.java [hunk 2]
```diff
@@ -98,6 +126,31 @@ public void deletePlayer(EPlayer ePlayer) {
...
     }
 
+    public void selectPlayer(UUID uuid, Consumer<EPlayer> callback) {
+        this.databaseConnector.connect(connection -> {
+            String selectPlayers = "SELECT * FROM " + this.getTablePrefix() + "players where uuid = ?";
+
+            try (PreparedStatement statement = connection.prepareStatement(selectPlayers)) {
+                statement.setString(1, uuid.toString());
+                ResultSet result = statement.executeQuery();
+                if (result.next()) {
```
Comment (SpraxDev) [inline src/main/java/com/songoda/epiclevels/database/DataManager.java:143]: Shouldn't the callback be invoked with `null` if we don't find a player?

File: src/main/java/com/songoda/epiclevels/database/DataUpdaterAbstract.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+            });
+        } catch (Throwable t) {
+            t.printStackTrace();
+        }
+        lock.readLock().unlock();
+    }
+
+    public void cleanMessages() {
+        lock.readLock().lock();
+        if (!isEnabled()) {
```
Comment (SpraxDev) [inline src/main/java/com/songoda/epiclevels/database/DataUpdaterAbstract.java:109]: Same thing about locking as a bit further up

File: src/main/java/com/songoda/epiclevels/database/DataUpdaterAbstract.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        } catch (Throwable t) {
+            t.printStackTrace();
+        }
+        lock.readLock().unlock();
+    }
+
+    public void getMessages(Consumer<String> callback) {
+        lock.readLock().lock();
+        if (!isEnabled()) {
+            lock.readLock().unlock();
```
Comment (SpraxDev) [inline src/main/java/com/songoda/epiclevels/database/DataUpdaterAbstract.java]: Same thing about locking as a bit further up

File: src/main/java/com/songoda/epiclevels/database/DataUpdaterAbstract.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    public boolean isEnabled() {
+        return enabled;
+    }
+
+    public void setEnabled(boolean enabled) {
+        this.enabled = enabled;
+    }
+
+    public void sendMessage(String msg) {
+        lock.readLock().lock();
```
Comment (SpraxDev) [inline src/main/java/com/songoda/epiclevels/database/DataUpdaterAbstract.java]: `#isEnabled()` just returns a boolean, could we avoid the lock and return instantly if it returns `false`?
(As `#setEnabled(boolean)` does not lock itself I don't see any benefit.

File: src/main/java/com/songoda/epiclevels/database/DataUpdater.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                break;
+            default:
+                throw new IllegalArgumentException("Cannot process update with the ID: " + id);
+        }
+    }
+
+    public void processPlayerUpdate(UUID uuid) {
+        if (toUpdate.contains(uuid)) {
+            return;
+        } else {
```
Comment (SpraxDev) [inline src/main/java/com/songoda/epiclevels/database/DataUpdater.java]: If we return inside the if we can get rid of the `else`-Block for better readability

**Review by SpraxDev — CHANGES_REQUESTED** [review]
Hi, I've tried reviewing your PR to my best ability:

1. Could you please direct your PR do the `development` branch? If there are conflicts or something ignore this request and I'll deal with them this time - We don't mention this anywhere right now 
2. I'm not really sure about the new `messenger` table - What's the use-case it's good for?

And I must say I'm certainty not happy about Exceptions just being printed to the console :/
It's better than just hiding them but in the future we really should improve this.

**Comment by Rubenicos** [comment]
> Hi, I've tried reviewing your PR to my best ability:
> 
> 1. Could you please direct your PR do the `development` branch? If there are conflicts or something ignore this request and I'll deal with them this time - We don't mention this anywhere right now

Done

> 2. I'm not really sure about the new `messenger` table - What's the use-case it's good for?

It's a new table only for communication purposes, all the hosts connected to the same database will receive update messages via `messenger` table to update cached player data (or boosts) in runtime without having the player online, then the rows are deleted.
In other words, for multi-host compatibility.

> And I must say I'm certainty not happy about Exceptions just being printed to the console :/ It's better than just hiding them but in the future we really should improve this.

Sure, actually I put that because it's almost impossible to get an error in those operations, but if you want i will remove it

**Comment by SpraxDev** [comment]
> Sure, actually I put that because it's almost impossible to get an error in those operations, but if you want i will remove it

No I think you misunderstood - Please keep them in.
I think `#printStacktrace()` is better than not having it. What I don't like about it is that 90 % of our users do not or are unable to read the error logs. So we will never see them. Depending on when the error happens and what the consequences are, some of our plugins just get more broken over time because it is in an unexpected state.

I personally prefer for methods of this kind to throw the Exceptions and handle it somewhere higher in the *'list'* but I didn't fully read your code so maybe you already are at that high point I mean.

**Review by SpraxDev — APPROVED** [review]
Looks good to me - Thanks for your contribution!

**Pull request merged**